dm/worker/server.go

```diff
@@ -163,6 +159,59 @@ func (s *Server) Start() error {
 	return terror.ErrWorkerStartService.Delegate(err)
 }
 
+func (s *Server) observeSourceBound(ctx context.Context, etcdCli *clientv3.Client, rev int64) error {
+	var wg sync.WaitGroup
+	for {
+		sourceBoundCh := make(chan ha.SourceBound, 10)
+		sourceBoundErrCh := make(chan error, 10)
+		wg.Add(1)
+		// use ctx1, cancel1 to make sure old watcher has been released
+		ctx1, cancel1 := context.WithCancel(ctx)
+		go func() {
+			defer func() {
+				close(sourceBoundCh)
+				close(sourceBoundErrCh)
+				wg.Done()
+			}()
+			ha.WatchSourceBound(ctx1, etcdCli, s.cfg.Name, rev+1, sourceBoundCh, sourceBoundErrCh)
+		}()
+		err := s.handleSourceBound(ctx1, sourceBoundCh, sourceBoundErrCh)
+		cancel1()
+		wg.Wait()
+
+		if etcdutil.IsRetryableError(err) {
+			rev = 0
+			retryNum := 1
+			for rev == 0 {
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-time.After(500 * time.Millisecond):
+					s.stopWorker("")
```

Comment:
Now, we call `s.stopWorker("")` both for keep alive interupt and etcd compact? Is this necessary?

Reply:
revised in https://github.com/pingcap/dm/pull/499/commits/709ef7bfc344c5b1d5cf33dcc9f559112300382e